### DISCUSSION TITLE
cudaFree(0) from Emgu ?

### DISCUSSION THREAD

**Discussion by observer11111** [emgucv/emgucv#405] [Q&A]
Is there a way to call cudaFree(0) from Emgu?  

Calling cudaFree(0) prior to the timed portion of the application will trigger CUDA context creation and the associated overhead. I'd like to be able to do that from Emgu.

**Comment by emgucv** [comment]
It is not a feature of Open CV and we do not have plan to add it in the short term.

We have also looked into the documentation for cudaFree:
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1ga042655cbbf3408f01061652a075e094

> __host__​__device__​cudaError_t cudaFree ( void* devPtr )
> Frees memory on the device.
> Description
> ... If devPtr is 0, no operation is performed. ...

It looks like it will do nothing if 0 is passed as devPtr.

Btw, calling `CudaInvoke.HasCuda` should also trigger CUDA context creation and the associated overhead. Please give it a try.

**Reply by observer11111** [reply]
Thanks for the suggestion: I'll try that.   I do use CudaInvoke.HasCuda elsewhere in the program, but I'll try using it immediately before the CUDA operations which seems to benefit from a warm-up (CaffeNet Net forward operation using cuda backend option).    The cudaFree(0) idea I see mentioned occasionally, like here https://forums.developer.nvidia.com/t/very-fast-ramp-down-from-high-to-low-clock-speeds-leading-to-increased-time-repeatedly-ramping-up/165974/10 so I am surprised to see that it would do nothing at all.
